File: leetcode/37.sudoku-solver/cpp/solvesudoku.cpp
#include <vector>

const int SIZE = 9;

struct point
{
    int row;
    int col;
};

class Solution
{
public:
    void solveSudoku(std::vector<std::vector<char>> &board)
    {
        board_ = board;
        backtrack(0);
        board = board_;
    }

private:
    std::vector<std::vector<char>> board_{};

    point idxToPoint(int idx)
    {
        return {idx / SIZE, idx % SIZE};
    }

    bool isValid(int row, int col, char v)
    {
        for (int i = 0; i < SIZE; ++i)
        {
            if (board_[row][i] == v)
                return false;

            if (board_[i][col] == v)
                return false;
        }

        int startRow = (row / 3) * 3;
        int startCol = (col / 3) * 3;

        for (int r = startRow; r < startRow + 3; ++r)
        {
            for (int c = startCol; c < startCol + 3; ++c)
            {
                if (board_[r][c] == v)
                    return false;
            }
        }

        return true;
    }

    bool backtrack(int start)
    {
        if (start == SIZE * SIZE)
            return true;

        auto point = idxToPoint(start);

        if (board_[point.row][point.col] != '.')
            return backtrack(start + 1);

        for (auto i = 1; i < 10; i++)
        {
            if (!isValid(point.row, point.col, i + '0'))
                continue;

            board_[point.row][point.col] = i + '0';
            if (backtrack(start + 1))
                return true;
            board_[point.row][point.col] = '.';
        }

        return false;
    }
};

File: leetcode/37.sudoku-solver/ts/solvesudoku.ts
const SIZE = 9

const solveSudoku = (board: string[][]): void => {
    backtrack(board, 0);
}

const backtrack = (board: string[][], start: number): boolean => {
    if (start === SIZE * SIZE) return true

    const [row, col] = idxToPoint(start)

    if (board[row][col] !== ".") return backtrack(board, start + 1)

    for (let i = 1; i < 10; i++) {
        if (!isValid(board, row, col, i.toString())) continue

        board[row][col] = i.toString()

        if (backtrack(board, start + 1)) return true

        board[row][col] = "."
    }

    return false;
}

const isValid = (board: string[][], row: number, col: number, v: string): boolean => {
    for (let i = 0; i < SIZE; i++) {
        if (board[row][i] === v) return false
        if (board[i][col] === v) return false
    }

    const startRow = Math.floor(row / 3) * 3
    const startCol = Math.floor(col / 3) * 3

    for (let r = startRow; r < startRow + 3; r++) {
        for (let c = startCol; c < startCol + 3; c++) {
            if (board[r][c] === v) return false
        }
    }

    return true
}

const idxToPoint = (idx: number): [number, number] => {
    return [Math.floor(idx / SIZE), idx % SIZE]
}

export default solveSudoku
